### PR TITLE
Gradual Variable Interpolation Process Implementation

### PR DESCRIPTION
**Pull request by SADPR**
# Gradual Variable Interpolation Process Implementation

 _**Please ignore the name of the branch.**_

This pull request introduces a new process to the MeshingApplication for gradually interpolating nodal values from an origin model part to a destination model part. The rate of interpolation can be controlled through an `alpha_increment` parameter or by specifying the total number of steps for the full interpolation `steps_for_full_interpolation`.

## Scope

The implemented process supports a user-defined variables list for interpolation, providing the flexibility to apply it to different applications.

## Application

The new process will be particularly useful for applications where gradual data transfer between different model parts is required, such as in multi-physics simulations.

## Changes

- New `gradual_variable_interpolation_process.py` added to the MeshingApplication's python scripts.
- Accompanying unit test `test_gradual_variable_interpolation_process.py` is added to the MeshingApplication's tests.
- Test model parts and expected outputs files are added to a new `gradual_variable_interpolation_test_files` directory within MeshingApplication's tests.

## Note

This implementation is part of a deliverable that we aim to fulfill in the near future, hence its swift integration into the master is beneficial.

## Files Added

- applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py
- applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/ExpectedOutput.npy
- applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/destination_model_part.mdpa
- applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/origin_model_part.mdpa
- applications/MeshingApplication/tests/test_gradual_variable_interpolation_process.py

### CODE REVIEW

File: applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/ExpectedOutput.npy
Comment (loumalouomega) [inline applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/ExpectedOutput.npy:1]: What is this?

File: applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import KratosMultiphysics
+import KratosMultiphysics.MeshingApplication as KratosMA
+
+def Factory(settings, model):
+    if not isinstance(settings, KratosMultiphysics.Parameters):
+        raise Exception("expected input shall be a Parameters object, encapsulating a json string")
+    return GradualVariableInterpolationProcess(model, settings["Parameters"])
+
+class GradualVariableInterpolationProcess(KratosMultiphysics.Process):
```
Comment (loumalouomega) [inline applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py:9]: Why not C¿?

**Review by loumalouomega — COMMENTED** [review]
Comments

File: applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/ExpectedOutput.npy
Comment (SADPR) [reply applications/MeshingApplication/tests/gradual_variable_interpolation_test_files/ExpectedOutput.npy:1]: This is the expected result.

File: applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py (hunk 1 above)
Comment (SADPR) [reply applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py:9]: Due to tight project deadlines, we prioritized a quick Python implementation. However, migrating this to C++ for better performance is on our to-do list.

File: applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        destination_model_part_name = self.settings["destination_model_part_name"].GetString()
+        self.destination_model_part = self.model.GetModelPart(destination_model_part_name)
+
+        self.InterpolateVariables()
+
+    def InterpolateVariables(self):
+        #Import Origin Model Part 
+        self.origin_model_part = self.model.CreateModelPart("OriginModelPart")
+        for variable in self.destination_model_part.GetHistoricalVariablesNames():
+            if KratosMultiphysics.KratosGlobals.HasVariable(variable):
```
Comment (matekelemen) [inline applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py:61]: should throw on the `else` branch

File: applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                self.origin_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.KratosGlobals.GetVariable(variable))
+
+        model_part_io = KratosMultiphysics.ModelPartIO(self.origin_model_part_file_name)
+        model_part_io.ReadModelPart(self.origin_model_part) 
+
+        #Set velocity field to origin model part
+        for node in self.origin_model_part.Nodes:
+            for variable in self.interpolation_variables_list:
+                node.SetValue(variable, self.alpha_increment * node.GetSolutionStepValue(variable))
```
Comment (matekelemen) [inline applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py]: this loop should be in C++

File: applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        #Interpolate velocity to destination model part
```
Comment (matekelemen) [inline applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py]: velocity?

File: applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        interpolation.Execute()
+        if self.constrain_variables:
+            for node in self.destination_model_part.Nodes:
+                for variable in self.interpolation_variables_list:
+                    node.SetSolutionStepValue(variable, node.GetValue(variable))
+                    node.Fix(variable)
+        else:
+            for node in self.destination_model_part.Nodes:
+                for variable in self.interpolation_variables_list:
+                    node.SetSolutionStepValue(variable, node.GetValue(variable))
```
Comment (matekelemen) [inline applications/MeshingApplication/python_scripts/gradual_variable_interpolation_process.py]: again, these loops should be in C++

**Review by matekelemen — COMMENTED** [review]
Apart from the heavy loops being in python, my problem is that this process is a combination of three others that (I believe) are already implemented.
1) read a `ModelPart` from an `.mdpa` file
2) mapping nodal variables between `ModelPart`s
3) scale nodal variables (similar to rampups in the fluid dynamics app)

Am I missing something?

**Comment by SADPR** [comment]
I've updated the pull request following the suggestions made by @matekelemen and @loumalouomega. In response to these, I created 'gradual_variable_interpolation_utility.h', which allowed me to move the loop structures into the C++ side and utilize parallel utilities for improved performance. I've also addressed the 'velocities' comments for clarity and added the recommended exception handling. Looking forward to any additional feedback.

**Comment by loumalouomega** [comment]
Can be separated in header and source files?

**Comment by SADPR** [comment]
The code is now organized into header and source files, @loumalouomega. I believe it's ready for a final review and subsequent merge.

File: applications/MeshingApplication/custom_utilities/gradual_variable_interpolation_utility.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
+#ifndef GRADUAL_VARIABLE_INTERPOLATION_UTILITY_H
```
Comment (loumalouomega) [inline applications/MeshingApplication/custom_utilities/gradual_variable_interpolation_utility.h]: Use Pragma once

**Review by loumalouomega — COMMENTED** [review]
Minor comment

File: applications/MeshingApplication/custom_utilities/gradual_variable_interpolation_utility.h (hunk 6 above)
Comment (SADPR) [reply applications/MeshingApplication/custom_utilities/gradual_variable_interpolation_utility.h]: Done @loumalouomega.

**Review by loumalouomega — APPROVED** [review]
Okay from my side

**Pull request merged**